src/dune_rules/scope.ml

```diff
@@ -25,78 +25,110 @@ module DB = struct
 
   module Found_or_redirect : sig
     type t = private
-      | Found of Lib_info.external_
-      | Redirect of (Loc.t * Lib_name.t)
+      | Found of (Toggle.t * Lib_info.external_)
+      | Redirect of (Toggle.t * (Loc.t * Lib_name.t))
 
-    val redirect : Lib_name.t -> Loc.t * Lib_name.t -> Lib_name.t * t
-    val found : Lib_info.external_ -> t
+    val redirect : enabled:Toggle.t -> Lib_name.t -> Loc.t * Lib_name.t -> Lib_name.t * t
+    val found : enabled:Toggle.t -> Lib_info.external_ -> t
   end = struct
     type t =
-      | Found of Lib_info.external_
-      | Redirect of (Loc.t * Lib_name.t)
+      | Found of (Toggle.t * Lib_info.external_)
+      | Redirect of (Toggle.t * (Loc.t * Lib_name.t))
 
-    let redirect from (loc, to_) =
+    let redirect ~enabled from (loc, to_) =
       if Lib_name.equal from to_
       then Code_error.raise ~loc "Invalid redirect" [ "to_", Lib_name.to_dyn to_ ]
-      else from, Redirect (loc, to_)
+      else from, Redirect (enabled, (loc, to_))
     ;;
 
-    let found x = Found x
+    let found ~enabled x = Found (enabled, x)
   end
 
   module Library_related_stanza = struct
     type t =
       | Library of Path.Build.t * Library.t
-      | Library_redirect of Library_redirect.Local.t
+      | Library_redirect of Path.Build.t * Library_redirect.Local.t
       | Deprecated_library_name of Deprecated_library_name.t
   end
 
   let create_db_from_stanzas ~instrument_with ~parent ~lib_config stanzas =
-    let map =
+    let+ map =
       List.map stanzas ~f:(fun stanza ->
         match (stanza : Library_related_stanza.t) with
-        | Library_redirect s ->
-          let old_public_name = Lib_name.of_local s.old_name in
-          Found_or_redirect.redirect old_public_name s.new_public_name
+        | Library_redirect (dir, s) ->
+          let old_public_name = Lib_name.of_local s.old_name.name in
+          let info =
+            let expander = Expander0.get ~dir in
+            Library.to_lib_info s.old_name ~expander ~dir ~lib_config |> Lib_info.of_local
+          in
+          let* enabled =
```

Comment:
The problem with this approach is that you're now evaluating the `enabled_if` so early, that you will get a cycle anytime you use a non trivial expression in `enabled_if`. Sure that works if you just select based on the context, but if you do something like:

```
(library
 (name foo)
 (enabled_if %{read:foo}))
(rule (with-stdout-to foo (echo true))
```
You will now get a cycle

Reply:
Thanks, I certainly didn't think about this case. I understand it's a quite edgy case, as it was not possible to use `read` variable in `enabled_if` until a few days ago when you added support for that.

I opened https://github.com/ocaml/dune/pull/10298 to understand what the current behavior is in `main`. Then I ran the same test in this branch. The cycle is identified correctly here too, but there is some difference in the report message, with some lines missing in this branch:

```diff
  Error: Dependency cycle between:
-    library "foo" in _build/default
  -> %{read:foo} at dune:3
- -> library "foo" in _build/default
```

So in this branch, the message doesn't include the `library` lines, but users can still see there's a cycle and get pointed to the precise file and line where it's happening.

All things considered, my opinion is that the approach used in this PR is a lot simpler than the other one, and introducing a lot of complexity to keep this error message is not worth it.

Reply:
> Then I ran the same test in this branch. 

See https://github.com/ocaml/dune/pull/10296/commits/815cc56a86d126cb0d2c4962610c3c6c8bd4e697.

Reply:
Error messages are not the point and this isn't an edge case. The final goal is to lift such limitations from enabled_if, and this PR is going the other way in that regard.